File: backend/app/agents/takeaway_extractor.py
"""
AI agent for extracting key takeaways from podcast transcripts.
Identifies and formats the most important insights and actionable points.
"""

import json
from typing import Dict, Any, List

from app.agents.base_agent import BaseAgent, AgentProcessingError
from app.utils.logger import get_logger

logger = get_logger(__name__)


class TakeawayExtractorAgent(BaseAgent):
    """
    AI agent that extracts key takeaways and insights from podcast transcripts.
    
    Identifies the most important, actionable, and memorable points from the
    discussion that would be valuable for someone to remember or act upon.
    """

    def process(self, transcript_content: str, summary: str = None, **kwargs) -> Dict[str, Any]:
        """
        Extract key takeaways from the podcast transcript.
        
        Args:
            transcript_content: The raw transcript text to analyze
            summary: Optional summary to provide context (from summarizer agent)
            **kwargs: Additional parameters (unused)
            
        Returns:
            Dictionary with 'takeaways' key containing list of key insights
            
        Raises:
            AgentProcessingError: If takeaway extraction fails
        """
        logger.info(f"[{self.agent_name}] Starting extraction of key points",
                   transcript_length=len(transcript_content),
                   has_summary=bool(summary))

        if not transcript_content.strip():
            raise AgentProcessingError("Cannot extract takeaways from empty transcript")

        # Create the extraction prompt
        prompt = self._build_extraction_prompt(transcript_content, summary)
        
        # Create system prompt for the extractor
        system_prompt = self._build_system_prompt()

        try:
            # Get takeaways from Claude
            raw_response = self._call_claude(prompt, system_prompt)
            
            # Parse and validate the takeaways
            takeaways = self._parse_takeaways(raw_response)
            
            if not takeaways:
                raise AgentProcessingError("No takeaways extracted from transcript")

            logger.info(f"[{self.agent_name}] Found {len(takeaways)} key takeaways")
            
            # Log the takeaways for visibility
            for i, takeaway in enumerate(takeaways, 1):
                logger.info(f"[{self.agent_name}]   {i}. {self._truncate_for_log(takeaway, 150)}")

            logger.info(f"[{self.agent_name}] Complete")

            return {"takeaways": takeaways}

        except Exception as e:
            logger.error(f"[{self.agent_name}] Failed to extract takeaways", error=str(e))
            raise AgentProcessingError(f"Takeaway extraction failed: {str(e)}")

    def _build_system_prompt(self) -> str:
        """Build the system prompt for Claude."""
        return """You are an expert at identifying key insights and actionable takeaways from podcast discussions.

Your task is to extract the most important, valuable, and memorable points that:
- Represent key insights or learnings shared during the discussion
- Are actionable or applicable to the audience
- Capture important facts, statistics, or expert opinions
- Highlight notable quotes or profound statements
- Include practical advice or recommendations mentioned
- Cover significant predictions or future outlook discussed

Focus on substantive content that would be valuable for someone to remember or act upon. Avoid:
- Basic introductory statements
- Small talk or casual conversation
- Obvious or common knowledge points
- Repetitive information

Return your response as a simple numbered list, with each takeaway as a complete, clear sentence."""

    def _build_extraction_prompt(self, transcript_content: str, summary: str = None) -> str:
        """
        Build the prompt for extracting takeaways.
        
        Args:
            transcript_content: The transcript to analyze
            summary: Optional summary for context
            
        Returns:
            Formatted prompt string
        """
        # Truncate very long transcripts for the prompt
        max_transcript_length = 12000  # Reasonable limit for Claude context
        if len(transcript_content) > max_transcript_length:
            transcript_content = transcript_content[:max_transcript_length] + "\n[...transcript truncated...]"
            logger.info(f"[{self.agent_name}] Truncated long transcript for processing")

        prompt = f"""Analyze the following podcast transcript and extract the key takeaways and insights.

Focus on identifying:
- Important facts, statistics, or expert insights
- Actionable advice or recommendations
- Significant predictions or future outlook
- Notable quotes or profound statements
- Key lessons learned or wisdom shared
- Practical tips mentioned

"""

        # Add summary context if available
        if summary:
            prompt += f"""CONTEXT SUMMARY:
{summary}

"""

        prompt += f"""TRANSCRIPT:
{transcript_content}

Please extract 4-8 key takeaways from this podcast. Format your response as a simple numbered list:

1. [First key takeaway]
2. [Second key takeaway]
3. [Third key takeaway]
etc.

KEY TAKEAWAYS:"""

        return prompt

    def _parse_takeaways(self, raw_response: str) -> List[str]:
        """
        Parse the takeaways from Claude's response.
        
        Args:
            raw_response: Raw response text from Claude
            
        Returns:
            List of cleaned takeaway strings
        """
        takeaways = []
        
        # Split response into lines
        lines = raw_response.strip().split('\n')
        
        for line in lines:
            line = line.strip()
            if not line:
                continue
            
            # Look for numbered list items
            # Patterns: "1. text", "1) text", "- text", "• text"
            import re
            
            # Remove list markers
            patterns = [
                r'^\d+\.\s*',     # 1. 
                r'^\d+\)\s*',     # 1) 
                r'^-\s*',         # - 
                r'^•\s*',         # • 
                r'^\*\s*',        # * 
            ]
            
            cleaned_line = line
            for pattern in patterns:
                cleaned_line = re.sub(pattern, '', cleaned_line)
            
            # Skip if line is too short or doesn't look like a takeaway
            if len(cleaned_line.split()) < 3:
                continue
            
            # Skip common non-takeaway phrases
            skip_phrases = [
                "key takeaways",
                "takeaways:",
                "summary:",
                "in conclusion",
                "to summarize"
            ]
            
            if any(phrase in cleaned_line.lower() for phrase in skip_phrases):
                continue
            
            # Ensure it ends with proper punctuation
            if cleaned_line and not cleaned_line.endswith(('.', '!', '?')):
                cleaned_line += '.'
            
            # Ensure it starts with capital letter
            if cleaned_line and not cleaned_line[0].isupper():
                cleaned_line = cleaned_line[0].upper() + cleaned_line[1:]
            
            if cleaned_line:
                takeaways.append(cleaned_line)
        
        # Limit to reasonable number of takeaways
        if len(takeaways) > 10:
            takeaways = takeaways[:10]
            logger.warning(f"[{self.agent_name}] Truncated takeaways list to 10 items")
        
        return takeaways

File: internal/agents/takeaway_extractor.go
package agents

import (
	"context"
	"regexp"
	"strings"
	"time"
	
	"podcast-analyzer/internal/clients"
	"podcast-analyzer/internal/config"
)

// TakeawayExtractorAgent extracts key takeaways and insights from podcast transcripts
type TakeawayExtractorAgent struct {
	*BaseAgent
	anthropicClient clients.AnthropicClientInterface
}

// NewTakeawayExtractorAgent creates a new takeaway extractor agent
func NewTakeawayExtractorAgent(cfg *config.Config) *TakeawayExtractorAgent {
	return &TakeawayExtractorAgent{
		BaseAgent:       NewBaseAgent("takeaway_extractor"),
		anthropicClient: clients.NewAnthropicClient(cfg),
	}
}

// Process extracts key takeaways from the podcast transcript
func (t *TakeawayExtractorAgent) Process(ctx context.Context, content string) (Result, error) {
	return t.ProcessWithOptions(ctx, content, ProcessingOptions{})
}

// ProcessWithOptions extracts key takeaways with optional summary context
func (t *TakeawayExtractorAgent) ProcessWithOptions(ctx context.Context, content string, opts ProcessingOptions) (Result, error) {
	start := time.Now()
	defer func() {
		t.LogAPICall(ctx, "anthropic", len(content), true)
	}()
	
	// Log start of processing
	t.LogStart(ctx, len(content))
	
	// Validate content
	if err := t.ValidateContent(content); err != nil {
		t.LogError(ctx, err, time.Since(start))
		return Result{}, err
	}
	
	// Build prompts
	systemPrompt := t.buildSystemPrompt()
	userPrompt := t.buildUserPrompt(content, opts.Summary)
	
	// Call Claude API
	rawResponse, err := t.anthropicClient.CallClaude(ctx, t.Name(), userPrompt, systemPrompt, false)
	if err != nil {
		t.LogError(ctx, err, time.Since(start))
		return Result{}, NewAgentError(t.Name(), "failed to extract takeaways", err)
	}
	
	// Parse and validate the takeaways
	takeaways := t.parseTakeaways(rawResponse)
	if len(takeaways) == 0 {
		err := NewAgentError(t.Name(), "no takeaways extracted from transcript", nil)
		t.LogError(ctx, err, time.Since(start))
		return Result{}, err
	}
	
	result := Result{Takeaways: takeaways}
	
	// Log success with takeaway details
	t.logTakeaways(ctx, takeaways)
	t.LogSuccess(ctx, &result, time.Since(start))
	
	return result, nil
}

// buildSystemPrompt creates the system prompt for Claude
func (t *TakeawayExtractorAgent) buildSystemPrompt() string {
	return `You are an expert at identifying key insights and actionable takeaways from podcast discussions.

Your task is to extract the most important, valuable, and memorable points that:
- Represent key insights or learnings shared during the discussion
- Are actionable or applicable to the audience
- Capture important facts, statistics, or expert opinions
- Highlight notable quotes or profound statements
- Include practical advice or recommendations mentioned
- Cover significant predictions or future outlook discussed

Focus on substantive content that would be valuable for someone to remember or act upon. Avoid:
- Basic introductory statements
- Small talk or casual conversation
- Obvious or common knowledge points
- Repetitive information

Return your response as a simple numbered list, with each takeaway as a complete, clear sentence.`
}

// buildUserPrompt creates the user prompt with transcript and optional summary
func (t *TakeawayExtractorAgent) buildUserPrompt(content, summary string) string {
	// Truncate very long transcripts for the prompt
	maxTranscriptLength := 12000 // Reasonable limit for Claude context
	if len(content) > maxTranscriptLength {
		content = t.TruncateContent(content, maxTranscriptLength)
	}
	
	prompt := `Analyze the following podcast transcript and extract the key takeaways and insights.

Focus on identifying:
- Important facts, statistics, or expert insights
- Actionable advice or recommendations
- Significant predictions or future outlook
- Notable quotes or profound statements
- Key lessons learned or wisdom shared
- Practical tips mentioned

`
	
	// Add summary context if available
	if summary != "" {
		prompt += "CONTEXT SUMMARY:\n" + summary + "\n\n"
	}
	
	prompt += "TRANSCRIPT:\n" + content + "\n\n"
	prompt += `Please extract 4-8 key takeaways from this podcast. Format your response as a simple numbered list:

1. [First key takeaway]
2. [Second key takeaway]
3. [Third key takeaway]
etc.

KEY TAKEAWAYS:`
	
	return prompt
}

// parseTakeaways parses takeaways from Claude's response
func (t *TakeawayExtractorAgent) parseTakeaways(rawResponse string) []string {
	var takeaways []string
	
	// Split response into lines
	lines := strings.Split(strings.TrimSpace(rawResponse), "\n")
	
	for _, line := range lines {
		processedLine := t.processTakeawayLine(line)
		if processedLine != "" {
			takeaways = append(takeaways, processedLine)
		}
	}
	
	// Limit to reasonable number of takeaways
	if len(takeaways) > 10 {
		t.logger.WithFields(map[string]interface{}{
			"agent":            t.Name(),
			"original_count":   len(takeaways),
			"truncated_count":  10,
		}).Warn("Truncated takeaways list to maximum count")
		takeaways = takeaways[:10]
	}
	
	return takeaways
}

// removeListMarkers removes numbered and bulleted list markers from a line
func (t *TakeawayExtractorAgent) removeListMarkers(line string) string {
	patterns := []*regexp.Regexp{
		regexp.MustCompile(`^\d+\.\s*`),     // 1. 
		regexp.MustCompile(`^\d+\)\s*`),     // 1) 
		regexp.MustCompile(`^-\s*`),         // - 
		regexp.MustCompile(`^•\s*`),         // • 
		regexp.MustCompile(`^\*\s*`),        // * 
	}
	
	cleanedLine := line
	for _, pattern := range patterns {
		cleanedLine = pattern.ReplaceAllString(cleanedLine, "")
	}
	
	return cleanedLine
}

// shouldSkipLine determines if a line should be filtered out as a non-takeaway
func (t *TakeawayExtractorAgent) shouldSkipLine(line string) bool {
	// Skip if line is too short
	words := strings.Fields(line)
	if len(words) < 3 {
		return true
	}
	
	// Skip common non-takeaway phrases
	skipPhrases := []string{
		"key takeaways",
		"takeaways:",
		"summary:",
		"in conclusion",
		"to summarize",
	}
	
	lowerLine := strings.ToLower(line)
	for _, phrase := range skipPhrases {
		if strings.Contains(lowerLine, phrase) {
			return true
		}
	}
	
	return false
}

// processTakeawayLine applies list marker removal, validation, and cleaning to a line
func (t *TakeawayExtractorAgent) processTakeawayLine(line string) string {
	line = strings.TrimSpace(line)
	if line == "" {
		return ""
	}
	
	// Remove list markers
	cleanedLine := t.removeListMarkers(line)
	
	// Check if line should be skipped
	if t.shouldSkipLine(cleanedLine) {
		return ""
	}
	
	// Clean up the takeaway
	cleanedLine = t.cleanTakeaway(cleanedLine)
	
	return cleanedLine
}

// cleanTakeaway cleans and formats a single takeaway
func (t *TakeawayExtractorAgent) cleanTakeaway(takeaway string) string {
	// Trim whitespace
	cleaned := strings.TrimSpace(takeaway)
	
	// Ensure it ends with proper punctuation
	if len(cleaned) > 0 && !strings.HasSuffix(cleaned, ".") && !strings.HasSuffix(cleaned, "!") && !strings.HasSuffix(cleaned, "?") {
		cleaned += "."
	}
	
	// Ensure it starts with capital letter
	if len(cleaned) > 0 && !t.IsUpperCase(cleaned[0]) {
		cleaned = strings.ToUpper(string(cleaned[0])) + cleaned[1:]
	}
	
	return cleaned
}

// logTakeaways logs individual takeaways for visibility
func (t *TakeawayExtractorAgent) logTakeaways(ctx context.Context, takeaways []string) {
	correlationID := getCorrelationID(ctx)
	
	for i, takeaway := range takeaways {
		t.logger.WithFields(map[string]interface{}{
			"agent":          t.Name(),
			"correlation_id": correlationID,
			"takeaway_num":   i + 1,
			"takeaway":       t.TruncateForLog(takeaway, 150),
		}).Info("Extracted takeaway")
	}
}